File: denovim/denops/config/plugins/treesitter.ts
import type { Plug } from "https://deno.land/x/dvpm@1.0.0/mod.ts";

export const treesitter: Plug[] = [
  {
    url: "nvim-treesitter/nvim-treesitter",
    after: async ({ denops }) => {
      await denops.call(
        `luaeval`,
        `require("nvim-treesitter.configs").setup(_A.param)`,
        {
          param: {
            sync_install: false,
            ensure_installed: [
              "bash",
              "markdown",
              "markdown_inline",
              "make",
              "cmake",
              "toml",
              "yaml",
              "json",
              "vim",
              "lua",
              "typescript",
              "python",
              "scheme",
              "clojure",
              "java",
              "kotlin",
            ],
            highlight: {
              enable: true,
              additional_vim_regex_highlighting: false,
            },
          },
        },
      );
    },
  },
  {
    url: "yioneko/nvim-yati",
    dependencies: [{ url: "nvim-treesitter/nvim-treesitter" }],
    after: async ({ denops }) => {
      await denops.call(
        `luaeval`,
        `require("nvim-treesitter.configs").setup(_A.param)`,
        {
          param: {
            yati: {
              enable: true,
              default_lazy: true,
              default_fallback: "auto",
            },
            indent: {
              enable: false,
            },
          },
        },
      );
    },
  },
  {
    url: "nvim-treesitter/nvim-treesitter-context",
    dependencies: [{ url: "nvim-treesitter/nvim-treesitter" }],
    after: async ({ denops }) => {
      await denops.cmd(`lua require("treesitter-context").setup()`);
    },
  },
  {
    url: "HiPhish/nvim-ts-rainbow2",
    dependencies: [{ url: "nvim-treesitter/nvim-treesitter" }],
    after: async ({ denops }) => {
      await denops.call(
        `luaeval`,
        `require("nvim-treesitter.configs").setup(_A.param)`,
        {
          param: {
            rainbow: {
              enable: true,
              query: "rainbow-parens",
            },
          },
        },
      );
    },
  },
  //{
  //  url: "nvim-treesitter/nvim-treesitter-textobject",
  //  dependencies: [{ url: "nvim-treesitter/nvim-treesitter" }],
  //  after: async ({ denops }) => {
  //    await denops.call(
  //      `luaeval`,
  //      `require("nvim-treesitter.configs").setup(_A.param)`,
  //      {
  //        param: {
  //          textobjects: {
  //            select: {
  //              enable: true,
  //              lookahed: true
  //            },
  //          },
  //        },
  //      },
  //    );
  //  },
  //},
];

File: nvim/lua/rc/init/plugins/treesitter.lua
local spec = {
  {
    "nvim-treesitter/nvim-treesitter",
    lazy = true,
    event = { "BufReadPre", "BufNewFile" },
    build = function()
      require("nvim-treesitter.install").update({ with_sync = true })
    end,
    dependencies = {
      "nvim-treesitter/nvim-treesitter-context",
      "nvim-treesitter/nvim-treesitter-textobjects",
      "RRethy/nvim-treesitter-endwise",
      "andymass/vim-matchup",
    },
    opts = {
      highlight = {
        enable = true,
        disable = {},
      },
      indent = {
        enable = true,
        disable = { "scheme", "clojure", "python" },
      },
      ensure_installed = {
        -- shell
        "bash",
        -- documents
        "markdown",
        "markdown_inline",
        -- build tool
        "make",
        "cmake",
        -- conf file
        "toml",
        "yaml",
        "json",
        -- vim conf
        "vim",
        "lua",
        -- script
        "typescript",
        "python",
        -- lisp
        "scheme",
        "clojure",
        -- java
        "java",
        "kotlin",
        "groovy",
      },
      autotag = {
        enable = true,
      },
      endwise = {
        enable = true,
      },
      textobjects = {
        select = {
          enable = true,
          lookahead = true,
          keymaps = {
            ["af"] = "@function.outer",
            ["if"] = "@function.inner",
            ["ac"] = "@class.outer",
            ["ic"] = { query = "@class.inner", desc = "Select inner part of a class region" },
            ["as"] = { query = "@scope", query_group = "locals", desc = "Select language scope" },
          },

          selection_modes = {
            ["@parameter.outer"] = "v", -- charwise
            ["@function.outer"] = "V",  -- linewise
            ["@class.outer"] = "<c-v>", -- blockwise
          },
          include_surrounding_whitespace = true,
        },
        swap = {
          enable = true,
          swap_next = {
            ["<leader>a"] = "@parameter.inner",
          },
          swap_previous = {
            ["<leader>A"] = "@parameter.inner",
          },
        },
      },
      matchup = {
        enable = true,
      },
    },
    config = function(_, opts)
      require("nvim-treesitter.configs").setup(opts)
      vim.treesitter.language.register("typescript", "typescriptreact")
      vim.treesitter.language.register("bash", "zsh")
    end,
  },
  {
    "nvim-treesitter/nvim-treesitter-context",
    config = true,
  },
}

return spec
